File: examples/c++/context_menu.cpp
#include <string>
#include <vector>
#include <functional>

struct MenuItem {
    std::string label;
    std::string icon;
    bool enabled = true;
    bool separator = false;
    std::function<void()> onClick;
    void setEnabled(bool e) { enabled = e; }
    std::string render() { return (icon.empty() ? "" : icon + " ") + label; }
};

class ContextMenu {
public:
    int x = 0, y = 0;
    bool visible = false;
    std::vector<MenuItem> items;
    void show(int px, int py) { x = px; y = py; visible = true; }
    void hide() { visible = false; }
    void addItem(const MenuItem& item) { items.push_back(item); }
    void removeItem(const std::string& label) {
        items.erase(std::remove_if(items.begin(), items.end(),
            [&](const MenuItem& i) { return i.label == label; }), items.end());
    }
    MenuItem* hitTest(int mx, int my) {
        if (!visible) return nullptr;
        int cy = y;
        for (auto& item : items) {
            if (item.separator) { cy += 4; continue; }
            if (mx >= x && mx <= x + 160 && my >= cy && my <= cy + 24) {
                return item.enabled ? &item : nullptr;
            }
            cy += 24;
        }
        return nullptr;
    }
};

File: examples/python/context_menu.py
import html as _html


class MenuItem:
    def __init__(self, label="", icon=""):
        self.label = label
        self.icon = icon
        self.enabled = True
        self.separator = False
        self.on_click = None

    def set_enabled(self, e):
        self.enabled = e

    def render(self):
        prefix = (self.icon + " ") if self.icon else ""
        return prefix + self.label


class ContextMenu:
    def __init__(self):
        self.x = 0
        self.y = 0
        self.visible = False
        self.items = []

    def show(self, px, py):
        self.x = px
        self.y = py
        self.visible = True

    def hide(self):
        self.visible = False

    def add_item(self, item):
        self.items.append(item)

    def remove_item(self, label):
        self.items = [i for i in self.items if i.label != label]

    def hit_test(self, mx, my):
        if not self.visible:
            return None
        cy = self.y
        for item in self.items:
            if item.separator:
                cy += 4
                continue
            if self.x <= mx <= self.x + 160 and cy <= my <= cy + 24:
                return item if item.enabled else None
            cy += 24
        return None

    def render(self):
        if not self.items:
            return ""
        items_html = "".join(
            '<div class="ctx-sep"></div>' if it.separator
            else f'<div class="ctx-item" onclick="{it.on_click or ""}">{_html.escape(it.render())}</div>'
            for it in self.items
        )
        return f'''<div id="ctxMenu" class="ctx-menu">{items_html}</div>'''
